data-prepper-plugins/csv-processor/src/main/java/com/amazon/dataprepper/plugins/processor/csv/CSVProcessor.java

```diff
@@ -12,22 +12,88 @@
 @DataPrepperPlugin(name="csv", pluginType = Processor.class, pluginConfigurationType = CSVProcessorConfig.class)
 public class CSVProcessor extends AbstractProcessor<Record<Event>, Record<Event>> {
+    private final List<String> header;
+    private final String quote_char;
+    private final String delimiter;
+    private final String source_key;
+    private final String headerSourceKey;
+    private final boolean hasUserSpecifiedHeader;
+    private final boolean hasHeaderSource;
+    private final Boolean isDeleteHeaderAfterProcessing;
+
     @DataPrepperPluginConstructor
     public CSVProcessor(final PluginMetrics pluginMetrics, final CSVProcessorConfig config) {
         super(pluginMetrics);
-        throw new UnsupportedOperationException();
+        this.source_key = config.getSource();
+        this.header = config.getColumnNames();
+        this.quote_char = config.getQuoteCharacter();
+        this.delimiter = config.getDelimiter();
+        this.headerSourceKey = config.getColumnNamesSourceKey();
+        this.isDeleteHeaderAfterProcessing = config.isDeleteHeader();
+
+        this.hasUserSpecifiedHeader = (this.header != null);
+        this.hasHeaderSource = (this.headerSourceKey != null);
     }
 
     @Override
     public Collection<Record<Event>> doExecute(final Collection<Record<Event>> records) {
+        CsvMapper mapper = createCsvMapper();
+        CsvSchema schema = createCsvSchema();
+
+        for (Record<Event> record : records) {
+
+            Event event = record.getData();
+
+            String message = event.get(this.source_key, String.class);
+            boolean this_event_has_header_source = this.hasHeaderSource && event.containsKey(this.headerSourceKey);
+
+            try {
+                MappingIterator<List<String>> message_it = mapper.readerFor(List.class).with(schema).readValues(message);
+
+                // otherwise the message is empty
+                if (message_it.hasNextValue()) {
+                    List<String> row = message_it.nextValue();
+
+                    if (this_event_has_header_source) {
+                        String header_unprocessed = event.get(this.headerSourceKey, String.class);
+                        MappingIterator<List<String>> header_it = mapper.readerFor(List.class).with(schema).readValues(header_unprocessed);
+                        // if header is empty, behaves correctly since columns are autogenerated.
+                        List<String> header_from_event_source = header_it.nextValue();
+                        putDataInEvent(event, header_from_event_source, row);
+                    } else {
+                        if (this.hasUserSpecifiedHeader) {
+                            putDataInEvent(event, this.header, row);
+                        }
+                        else {
+                            List<String> emptyHeader = new ArrayList<String>();
+                            putDataInEvent(event, emptyHeader, row);
+                        }
```

Comment:
I think you could simplify this section and make it more readable by moving the header processing into a helper method. You could also use `else if` to get rid of the nested if statements. I'm thinking something like

```
private List<String> parseHeader(....) {
    if (this_event_has_header_source) {
        ....
        return header_it.nextValue();
    } else if (this.hasUserSpecifiedHeader) {
        return this.header;
    } else {
       return new ArrayList<String>();
    }
}
```


This section would then become 
```
final List<String> header = parseHeader(....);
putDataInEvent(event, header, row);
```

Reply:
That's much cleaner, thanks!